AttitudeManager/Tests/AMTest.cpp

```diff
@@ -0,0 +1,295 @@
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+#include "AM.hpp"
+#include "config.hpp"
+
+namespace AM {
+
+TEST(AttitudeManager, MotorInitializationAndOutput) {
+
+    // No motors
+    {
+        config::NUM_MOTORS = 0;
+        config::motors = new config::Motor_t[config::NUM_MOTORS];
+
+        AM::Flightmode *control_algorithm  = config::flightmodes[0].flightmodeConstructor();
+        AM::AttitudeManager am{control_algorithm};
+
+        // Non null references
+        EXPECT_THAT(am.motorReferences_[0], ::testing::NotNull());
+        EXPECT_THAT(am.motorReferences_[1], ::testing::NotNull());
+        EXPECT_THAT(am.motorReferences_[2], ::testing::NotNull());
+        EXPECT_THAT(am.motorReferences_[3], ::testing::NotNull());
+        EXPECT_THAT(am.motorReferences_[4], ::testing::NotNull());
+
+        // Shouldn't do anything
+        am.outputToMotor(config::throttle, 40);
+        am.outputToMotor(config::pitch, 30);
+        am.outputToMotor(config::roll, 20);
+        am.outputToMotor(config::yaw, 10);
+
+        delete control_algorithm;
+        delete[] config::motors;
+        control_algorithm = nullptr;
+        config::motors = nullptr;
+    }
+
+    // 5 motors of yaw type, some are inverted, *DO MIXED DRIVERS WHEN THERE ARE IMPLEMENATIONS OF MORE THAN ONE DRIVER*
+    {
+        config::NUM_MOTORS = 5;
+        config::motors = new config::Motor_t[config::NUM_MOTORS];
+
+        config::motors[0] =
+            config::Motor_t{   //Yaw servo motor
+                .axis = config::yaw,
+                .isInverted = false,
+                .driverConstructor = config::constructObject<MotorChannel, PWMChannel,
```

Comment:
My only problem with this is that as the driverConstructor returns a pointer to MotorChannel base class, I'm not able to access any of the variables or methods I created inside my MockMotor class for testing purposes (such as a way to display the value it would normally set in __HAL_TIM_SET_COMPARE)

Reply:
Otherwise it makes sense that this is the best way to do it and I would prefer doing it this way, just not sure how to solve that problem

Reply:
Also placing the mock alongside the other motor types makes it that I have to compile PWMChannel too even though I'm not using it, which gives me compiler errors for missing SystemCoreClock, __HAL_TIM_SET_PRESCALER, etc which I can just stub but feels like I shouldn't need to. Not sure if it would be an acceptable solution to just change cmakelists from "${MOTORCHANNEL_SRC}/*.cpp" to "${MOTORCHANNEL_SRC}/MockChannel.cpp"

Reply:
> My only problem with this is that as the driverConstructor returns a pointer to MotorChannel base class, I'm not able to access any of the variables or methods I created inside my MockMotor class for testing purposes (such as a way to display the value it would normally set in __HAL_TIM_SET_COMPARE)

Don't use the configs for testing your class here. Create an instance of a mock motor channel and pass it to the constructor of AM, as mentioned (dependency injection).

Reply:
> Also placing the mock alongside the other motor types makes it that I have to compile PWMChannel too even though I'm not using it, which gives me compiler errors for missing SystemCoreClock, __HAL_TIM_SET_PRESCALER, etc which I can just stub but feels like I shouldn't need to. Not sure if it would be an acceptable solution to just change cmakelists from "${MOTORCHANNEL_SRC}/*.cpp" to "${MOTORCHANNEL_SRC}/MockChannel.cpp"

I would suggest trying then to place all of the mocks in their own folder, and use the original includes then? You're right that we shouldn't need to compile these other files that aren't being used.